<a href="https://colab.research.google.com/github/kiojoel/Rock-vs-Mine-Prediction/blob/main/Rock_vs_Mine_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The goal of this project is to develop a machine learning model that can accurately classify objects detected by sonar signals as either rocks or mines. The model will be trained and evaluated using the SONAR dataset, which contains sonar signal data.


Importing Dependecies

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Data collection and Processing

In [ ]:
# Loading dataset to a pandas DataFrame
dataset = pd.read_csv('/content/sonar data.csv', header=None)

In [ ]:
dataset.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [ ]:
# size of the dataset (rows, columns)
dataset.shape

(208, 61)

In [ ]:
# Dataset Statistics
dataset.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [ ]:
dataset.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [ ]:
dataset[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

Mine(M) ---> 111

Rock(R) ---> 97



In [ ]:
# grouping the data according to the mean
dataset.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [ ]:
# Seperating data and label
X = dataset.drop(columns=60, axis=1)
Y = dataset[60]

In [ ]:
print(X)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

In [ ]:
print(Y)

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object


Spliting data into Training and Test data

In [ ]:
X_train, X_test, Y_train , Y_test = train_test_split(X,Y, test_size=0.1, stratify=Y, random_state=1)

In [ ]:
print(X_train.shape, X_test.shape, Y_train.shape , Y_test.shape )

(187, 60) (21, 60) (187,) (21,)


We have 187 train data and 21 test data.


**Model Training** with Logistic Regression

In [ ]:
model = LogisticRegression()

In [ ]:
# training the model with training data
model.fit(X_train, Y_train)

LogisticRegression()

Model Evaluation

In [ ]:
# accuracy on training data
train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(train_prediction, Y_train)

In [ ]:
print('Accuracy of training data: ', training_data_accuracy)

Accuracy of training data:  0.8342245989304813


In [ ]:
# accuracy on test data
test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(test_prediction, Y_test)

In [ ]:
print('Accuracy of test data: ', testing_data_accuracy)

Accuracy of test data:  0.7619047619047619


Making a Predictive System

Prediction with a Rock data

In [37]:
input_data = (0.0293,0.0644,0.0390,0.0173,0.0476,0.0816,0.0993,0.0315,0.0736,
              0.0860,0.0414,0.0472,0.0835,0.0938,0.1466,0.0809,0.1179,0.2179,
              0.3326,0.3258,0.2111,0.2302,0.3361,0.4259,0.4609,0.2606,0.0874,
              0.2862,0.5606,0.8344,0.8096,0.7250,0.8048,0.9435,1.0000,0.8960,
              0.5516,0.3037,0.2338,0.2382,0.3318,0.3821,0.1575,0.2228,0.1582,
              0.1433,0.1634,0.1133,0.0567,0.0133,0.0170,0.0035,0.0052,0.0083,
              0.0078,0.0075,0.0105,0.0160,0.0095,0.0011)

# converting input_data to a numpy array

input_data_array = np.asarray(input_data)

# reshape the array
input_data_reshaped = input_data_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)
print(prediction)

if prediction[0] == 'R':
  print('The object is a Rock')
else:
  print('The object is a Mine')

['R']
The object is a Rock


Prediction with a Mine data

In [ ]:
mine_input_data = np.array([0.0363,0.0478,0.0298,0.0210,0.1409,0.1916,0.1349,
                            0.1613,0.1703,0.1444,0.1989,0.2154,0.2863,0.3570,
                            0.3980,0.4359,0.5334,0.6304,0.6995,0.7435,0.8379,
                            0.8641,0.9014,0.9432,0.9536,1.0000,0.9547,0.9745,
                            0.8962,0.7196,0.5462,0.3156,0.2525,0.1969,0.2189,
                            0.1533,0.0711,0.1498,0.1755,0.2276,0.1322,0.1056,
                            0.1973,0.1692,0.1881,0.1177,0.0779,0.0495,0.0492,
                            0.0194,0.0250,0.0115,0.0190,0.0055,0.0096,0.0050,
                            0.0066,0.0114,0.0073,0.0033]).reshape(1,-1)
prediction = model.predict(mine_input_data)
print(prediction)

if prediction[0] == 'R':
  print('The object is a Rock')
else:
  print('The object is a Mine')

['M']
The object is a Mine
